### PR TITLE
Building for NS1.4

### PR DESCRIPTION
**Pull request by mayankbansal018**

### CODE REVIEW

File: src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollector.cs [hunk 1]
```diff
@@ -98,7 +94,7 @@ private static MethodInfo GetMethodInfoFromType(Type type, string funcName, Type
             MethodInfo methodInfo = null;
 
             var typeInfo = type.GetTypeInfo();
-            methodInfo = typeInfo?.GetMethod(funcName, argumentTypes);
+            methodInfo = typeInfo.AsType()?.GetMethod(funcName, argumentTypes);
```
Comment (Faizan2304) [inline src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollector.cs]: Is any specific reason why we moving typeInfo.GetMethod to Type.GetMethod

**Review by Faizan2304 — APPROVED** [review]

File: src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollector.cs (hunk 1 above)
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollector.cs]: In NS1.4 TypeInfo does not contain definition for GetMethod, or GetConstructor.
Although Type will continue to contain definition GetMethod, & GetConstructor, Will confirm with Immo that whether this is correct or not.

File: src/Microsoft.TestPlatform.Client/DesignMode/DesignModeClient.cs [hunk 2]
```diff
@@ -108,7 +110,7 @@ public void HandleParentProcessExit()
             this.Dispose();
 
             EqtTrace.Info("DesignModeClient: Parent process exited, Exiting myself..");
-            Environment.Exit(1);
+            new PlatformEnvironment().ExitCurrentProcess(1);
```
Comment (smadala) [inline src/Microsoft.TestPlatform.Client/DesignMode/DesignModeClient.cs]: Enable skipped test in Designmodeclient.

File: src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollector.cs [hunk 3]
```diff
@@ -97,17 +93,15 @@ private static MethodInfo GetMethodInfoFromType(Type type, string funcName, Type
         {
             MethodInfo methodInfo = null;
 
-            var typeInfo = type.GetTypeInfo();
-            methodInfo = typeInfo?.GetMethod(funcName, argumentTypes);
+            methodInfo = type?.GetMethod(funcName, argumentTypes);
```
Comment (smadala) [inline src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollector.cs]: Nit: remove MethodInfo declaration.

**Review by smadala — APPROVED** [review]

File: src/Microsoft.TestPlatform.Client/DesignMode/DesignModeClient.cs [hunk 4]
```diff
@@ -108,7 +110,9 @@ public void HandleParentProcessExit()
             this.Dispose();
 
             EqtTrace.Info("DesignModeClient: Parent process exited, Exiting myself..");
-            Environment.Exit(1);
+
+            // ToDo: Initialize this in .ctor, so that it is mockable
```
Comment (codito) [inline src/Microsoft.TestPlatform.Client/DesignMode/DesignModeClient.cs]: Please do this and add a test :)

File: src/Microsoft.TestPlatform.Client/DesignMode/DesignModeClient.cs [hunk 5]
```diff
@@ -108,7 +110,9 @@ public void HandleParentProcessExit()
...
+            new PlatformEnvironment().ExitCurrentProcess(1);
```
Comment (codito) [inline src/Microsoft.TestPlatform.Client/DesignMode/DesignModeClient.cs]: I think the API could be called `Exit`, it doesn't seem appropriate for Environment to ever expose an API to exit another process.

File: src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollector.cs (hunk 3 above)
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollector.cs]: Why null check?

File: src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollector.cs [hunk 6]
```diff
@@ -97,17 +93,15 @@ private static MethodInfo GetMethodInfoFromType(Type type, string funcName, Type
...
             return methodInfo;
         }
 
         private static object CreateObjectFromType(Type type)
         {
             object obj = null;
 
-            var typeInfo = type.GetTypeInfo();
-            var constructorInfo = typeInfo?.GetConstructor(Type.EmptyTypes);
+            var constructorInfo = type?.GetConstructor(Type.EmptyTypes);
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollector.cs]: Why null check?

File: src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollector.cs [hunk 7]
```diff
@@ -123,12 +117,7 @@ private Assembly LoadInProcDataCollectorExtension(string codeBase)
             Assembly assembly = null;
             try
             {
-#if NET451
-                assembly = Assembly.LoadFrom(codeBase);
-#else
-                assembly = AssemblyLoadContext.Default.LoadFromAssemblyPath(codeBase);
-#endif
-
+                assembly = new PlatformAssemblyLoadContext().LoadAssemblyFromPath(codeBase);
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollector.cs]: Can we make it testable and unit tests please?

File: src/Microsoft.TestPlatform.ObjectModel/Navigation/PortableSymbolReader.cs [hunk 8]
```diff
@@ -96,16 +94,12 @@ private void PopulateCacheForTypeAndMethodSymbols(string binaryPath)
                     // At this point, the assembly should be already loaded into the load context. We query for a reference to
                     // find the types and cache the symbol information. Let the loader follow default lookup order instead of
                     // forcing load from a specific path.
-#if NET451
-                    var asm = Assembly.Load(AssemblyName.GetAssemblyName(binaryPath));
-#else
-                    var asm = Assembly.Load(AssemblyLoadContext.GetAssemblyName(binaryPath));
-#endif
+                    var asm = Assembly.Load(new PlatformAssemblyLoadContext().GetAssemblyNameFromPath(binaryPath));
```
Comment (codito) [inline src/Microsoft.TestPlatform.ObjectModel/Navigation/PortableSymbolReader.cs]: Make it testable and unit tests please

File: src/Microsoft.TestPlatform.PlatformAbstractions/net451/Runtime/PlatformAssemblyExtension.cs [hunk 9]
```diff
@@ -0,0 +1,15 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.VisualStudio.TestPlatform.PlatformAbstractions
+{
+    using System.Reflection;
+
+    public static class PlatformAssemblyExtension
```
Comment (codito) [inline src/Microsoft.TestPlatform.PlatformAbstractions/net451/Runtime/PlatformAssemblyExtension.cs]: nit: `PlatformAssemblyExtensions`

File: src/Microsoft.TestPlatform.PlatformAbstractions/net451/Runtime/PlatformAssemblyExtension.cs [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+    {
+        public static string GetAssemblyLocation(this Assembly assembly)
```
Comment (codito) [inline src/Microsoft.TestPlatform.PlatformAbstractions/net451/Runtime/PlatformAssemblyExtension.cs]: Require docs for public API.

Comment (codito) [inline src/Microsoft.TestPlatform.PlatformAbstractions/net451/Runtime/PlatformAssemblyExtension.cs]: Please fix these in other implementations

File: src/Microsoft.TestPlatform.PlatformAbstractions/uap10.0/Runtime/PlatformAssemblyExtension.cs [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+namespace Microsoft.VisualStudio.TestPlatform.PlatformAbstractions
+{
+    using System.IO;
+    using System.Reflection;
+
+    public static class PlatformAssemblyExtension
+    {
+        public static string GetAssemblyLocation(this Assembly assembly)
+        {
+            return Directory.GetCurrentDirectory();
```
Comment (codito) [inline src/Microsoft.TestPlatform.PlatformAbstractions/uap10.0/Runtime/PlatformAssemblyExtension.cs]: Good to add a comment on why this is different for UAP.

File: src/Microsoft.TestPlatform.Utilities/CodeCoverageDataAttachmentsHandler.cs [hunk 12]
```diff
@@ -55,21 +53,15 @@ public ICollection<AttachmentSet> HandleDataCollectionAttachmentSets(ICollection
...
 
             try
             {
-                Assembly assembly = null;
-#if NET451
-                assembly = Assembly.LoadFrom(assemblyPath);
-#else
-                assembly = AssemblyLoadContext.Default.LoadFromAssemblyPath(assemblyPath);
-#endif
+                Assembly assembly = new PlatformAssemblyLoadContext().LoadAssemblyFromPath(assemblyPath);
```
Comment (codito) [inline src/Microsoft.TestPlatform.Utilities/CodeCoverageDataAttachmentsHandler.cs]: Make it testable and add unit tests please.

File: src/testhost.x86/DefaultEngineInvoker.cs [hunk 13]
```diff
@@ -62,14 +60,19 @@ public void Invoke(IDictionary<string, string> argsDictionary)
...
-                var processHelper = new ProcessHelper();
-                processHelper.SetExitCallback(
-                    parentProcessId,
-                    () =>
-                        {
-                            EqtTrace.Info("DefaultEngineInvoker: ParentProcess '{0}' Exited.", parentProcessId);
-                            Environment.Exit(1);
-                        });
+
+                // In remote scenario we cannot monitor parent process, so we expect user to pass parentProcessId as -1
```
Comment (codito) [inline src/testhost.x86/DefaultEngineInvoker.cs]: In that case, is it the runtime providers responsibility to clean up the stale test host processes? Should it be called out explicitly somewhere?

File: src/testhost.x86/Program.cs [hunk 14]
```diff
@@ -46,8 +45,9 @@ public static void Main(string[] args)
             }
         }
 
-        private static void Run(string[] args)
+        public static void Run(string[] args)
```
Comment (codito) [inline src/testhost.x86/Program.cs]: Request to add a comment to explain this.

File: src/testhost/testhost.csproj [hunk 15]
```diff
@@ -33,6 +33,11 @@
       <Version>4.3.0</Version>
     </PackageReference>
   </ItemGroup>
+  <ItemGroup Condition=" '$(TargetFramework)' == 'netstandard1.4' ">
```
Comment (codito) [inline src/testhost/testhost.csproj]: Why is this required? Just curious since we had faced several issues with TypeConverter earlier, and not sure if TypeConverter 4.3.0 is compatible with UAP10.0?

**Review by codito — CHANGES_REQUESTED** [review]

File: src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollector.cs (hunk 3 above)
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollector.cs]: resolved

File: src/Microsoft.TestPlatform.Client/DesignMode/DesignModeClient.cs (hunk 5 above)
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.Client/DesignMode/DesignModeClient.cs]: resolved

File: src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollector.cs (hunk 6 above)
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollector.cs]: resolved

File: src/Microsoft.TestPlatform.PlatformAbstractions/net451/Runtime/PlatformAssemblyExtension.cs (hunk 9 above)
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.PlatformAbstractions/net451/Runtime/PlatformAssemblyExtension.cs]: resolved

File: src/Microsoft.TestPlatform.PlatformAbstractions/net451/Runtime/PlatformAssemblyExtension.cs (hunk 10 above)
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.PlatformAbstractions/net451/Runtime/PlatformAssemblyExtension.cs]: resolved

File: src/Microsoft.TestPlatform.PlatformAbstractions/uap10.0/Runtime/PlatformAssemblyExtension.cs (hunk 11 above)
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.PlatformAbstractions/uap10.0/Runtime/PlatformAssemblyExtension.cs]: resolved

File: src/testhost.x86/DefaultEngineInvoker.cs (hunk 13 above)
Comment (mayankbansal018) [inline src/testhost.x86/DefaultEngineInvoker.cs]: The responsibility to clean up test runner lies with it's RuntimeProvider, but if vstest.console is dead then runtime provider won't get a chance to clean up, the best we can achieve here is safely exit the UWP app.

File: src/testhost.x86/Program.cs (hunk 14 above)
Comment (mayankbansal018) [inline src/testhost.x86/Program.cs]: resolved

File: src/Microsoft.TestPlatform.Client/DesignMode/DesignModeClient.cs (hunk 4 above)
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.Client/DesignMode/DesignModeClient.cs]: resolved

File: src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollector.cs (hunk 7 above)
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollector.cs]: resolved

**Comment by mayankbansal018** [comment]
@dotnet-bot Test this please
